File: Data Structures & Algorithms/merge-intervals/submission-0.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        merged = [intervals[0]]

        for i in range(1,len(intervals)):
            s2,e2 = intervals[i]
            s1,e1 = merged[-1]

            if s2<=e1:
                merged[-1][1] = max(e1,e2)
            else:
                merged.append([s2,e2])
        return merged

File: Data Structures & Algorithms/partition-labels/submission-0.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        //know a character's first and last appearance
        //-> becomes a range. then for each char in that range, makes sure same
        //-> merge intervals. sort by start time

        vector<int> last(26); //26 chars

        for (int i = 0; i < s.size(); i++){
            last[s[i] - 'a'] = i;
        }

        //go through string
        int start = 0, end = 0;
        vector<int> res;

        for (int i = 0; i < s.size(); i++){
            end = max(last[s[i]-'a'], end);
            if (i == end){ //range ends @ index
                res.push_back(end-start+1);
                start = end + 1; //starts next index
            }
        }   
        return res;

    }
};
